libcaf_net/caf/net/basp/worker.hpp

```diff
@@ -0,0 +1,118 @@
+/******************************************************************************
+ *                       ____    _    _____                                   *
+ *                      / ___|  / \  |  ___|    C++                           *
+ *                     | |     / _ \ | |_       Actor                         *
+ *                     | |___ / ___ \|  _|      Framework                     *
+ *                      \____/_/   \_|_|                                      *
+ *                                                                            *
+ * Copyright 2011-2019 Dominik Charousset                                     *
+ *                                                                            *
+ * Distributed under the terms and conditions of the BSD 3-Clause License or  *
+ * (at your option) under the terms and conditions of the Boost Software      *
+ * License 1.0. See accompanying files LICENSE and LICENSE_ALTERNATIVE.       *
+ *                                                                            *
+ * If you did not receive a copy of the license files, see                    *
+ * http://opensource.org/licenses/BSD-3-Clause and                            *
+ * http://www.boost.org/LICENSE_1_0.txt.                                      *
+ ******************************************************************************/
+
+#pragma once
+
+#include <atomic>
+#include <cstdint>
+#include <vector>
+
+#include "caf/config.hpp"
+#include "caf/detail/abstract_worker.hpp"
+#include "caf/detail/worker_hub.hpp"
+#include "caf/net/basp/header.hpp"
+#include "caf/net/basp/message_queue.hpp"
+#include "caf/net/basp/remote_message_handler.hpp"
+#include "caf/net/fwd.hpp"
+#include "caf/node_id.hpp"
+#include "caf/resumable.hpp"
+
+namespace caf {
+namespace net {
+namespace basp {
+
+/// Deserializes payloads for BASP messages asynchronously.
+class worker : public detail::abstract_worker,
+               public remote_message_handler<worker> {
+public:
+  // -- friends ----------------------------------------------------------------
+
+  friend remote_message_handler<worker>;
+
+  // -- member types -----------------------------------------------------------
+
+  using super = detail::abstract_worker;
+
+  using scheduler_type = scheduler::abstract_coordinator;
+
+  using buffer_type = std::vector<byte>;
+
+  using hub_type = detail::worker_hub<worker>;
+
+  // -- constructors, destructors, and assignment operators --------------------
+
+  /// Only the ::worker_hub has access to the construtor.
+  worker(hub_type& hub, message_queue& queue, proxy_registry& proxies);
+
+  ~worker() override;
+
+  // -- management -------------------------------------------------------------
+
+  void launch(const node_id& last_hop, const basp::header& hdr,
+              const buffer_type& payload);
```

Comment:
```suggestion
              const span<byte>& payload);
```

The implementation copies the content into `payload_`. There's no reason to require an actual buffer here.